Document required client changes for Rekor v2

haydentherapper opened the pull request:
Fixes #108

<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficient time for discussions to take place. Please use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

#### Summary
<!--
 Explain the **motivation** for making this change. What existing problem does the pull request solve? How can reviewers test this PR?
-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->

#### Documentation
<!--

Does this change require an update to documentation? How will users implement your new feature?

Please reference a PR within https://docs.sigstore.dev

-->

haydentherapper commented on CLIENTS.md:
```diff
@@ -0,0 +1,264 @@
+be included in the TrustedRoot distributed via TUF. Clients may request timestamps
+from other trusted timestamp authorities as well.
+
+## SigningConfig support
```
@jku I added some more info about the SigningConfig and motivation behind the changes, lemme know if you have more questions that I can clarify in this doc

loosebazooka commented on CLIENTS.md:
```diff
@@ -0,0 +1,264 @@
+Sigstore now operates a timestamp authority at `timestamp.sigstore.dev` and
+`timestamp.sigstage.dev` for staging, and the roots for these services will
+be included in the TrustedRoot distributed via TUF. Clients may request timestamps
+from other trusted timestamp authorities as well.
```
I assume we are expecting clients that use other "trusted timestamp" authorities to inject that information into the trusted root? Client code should not fetch the timestamp cert from the provider

loosebazooka commented on CLIENTS.md:
````diff
@@ -0,0 +1,266 @@
+The service implements one write API, `/api/v2/log/entries`. Example JSON request bodies
+below:
+
+```json
````
I haven't tested this, but `jsonc` appears to be a markdown rendering formation for json with comments.

loosebazooka commented on CLIENTS.md:
```diff
@@ -0,0 +1,266 @@
+## Rekor v2 API
+
+Rekor v2 supports HTTP or gRPC, like Fulcio. For Go, we have implemented a client
+already. For other languages, they can either use the OpenAPI docs or generated
+gRPC service proto, or create their own client.
```
```suggestion
already. For other languages, they can either use the [OpenAPI docs](https://github.com/sigstore/rekor-tiles/tree/main/docs/openapi), [gRPC service proto](https://github.com/sigstore/rekor-tiles/tree/main/api/proto), or create their own client.
```

loosebazooka commented on CLIENTS.md:
```diff
@@ -0,0 +1,266 @@
+
+Clients must verify that verification works with a TrustedRoot
+with multiple logs with overlapping validity windows. Confirm that
+the client will fetch the correct verification key from the TrustedRoot.
```
Fetch correct key by matching logid/validity window?

haydentherapper replied:
Added. I'd say we use log ID for now. We should later align log ID with how we're computing log IDs for checkpoints.

cmurphy commented on CLIENTS.md:
```diff
@@ -0,0 +1,266 @@
+            "content": "<base64 signature>",
+            "verifier": {
+                "publicKey": {
+                    "raw_bytes": "<base64 DER-encoded public key>"
```
```suggestion
                    "rawBytes": "<base64 DER-encoded public key>"
```

cmurphy commented on CLIENTS.md:
```diff
@@ -0,0 +1,266 @@
+            ]
+        },
+        "verifier": {
+            "x509Certificate": "<base64 DER-encoded certificate>", # May also be a public key
```
Would be clearer if the example was repeated like you are for hashedrekord, with the `publicKey` field instead of `x509Certificate`

cmurphy commented on CLIENTS.md:
```diff
@@ -0,0 +1,266 @@
+
+When monitoring the log searching for entries, the monitor will not request
+entries by index, but by tile. Monitors can choose to only fetch complete
+tiles or request partial tiles. It is recommended to request partial tiles,
```
It might not be clear to someone brand-new to this what a partial tile is or why you would want it. Maybe add some clarification that the latest tile is generally a partial tile.

haydentherapper commented:
Proposing we change how clients find the correct tlog to use to verify a proof, in https://github.com/sigstore/protobuf-specs/pull/629. Will wait for reviews before updating this doc.

jku reviewed (commented):
I think the one case where I'm not confident is where the v2 rekor instance is added to signing config:
* before any clients are updated this is straight forward: clients keep signing with v1, nothing breaks
* when the client A is upgraded to support v2, the docs seem to imply that it should then only use the v2 instance when the signingconfig includes one  (correct?)
* this leads to a client incompatibility if other clients are not able to verify rekor v2 at this point yet

Basically we may need to confirm verification support across the ecosystem before we make signingconfig changes where a rekor v2 instance is enabled

haydentherapper commented on CLIENTS.md:
```diff
@@ -0,0 +1,266 @@
+in a bundle. Clients no longer need to transform the Rekor response into a `TLE`
+message to store in the bundle.
+
+## Handling Longer Requests
+
+Clients need to increase request timeouts to at least 10 seconds.
+
+Rekor now batches uploads so that checkpoints are published less frequently.
+Additionally, we plan to support synchronous witnessing, where third-party
+witnesses independently verify the consistency of the log and Rekor provides
+co-signed checkpoints with each upload response.
```
I'll add the line about uploading in parallel, but the client shouldn't assume the same checkpoint for each, as Tessera might trigger a new checkpoint creation in the middle of the set of uploads. If a client needs to save space, it would need to request the latest checkpoint and compute inclusion proofs for all entries, but once we introduce synchronous witnessing, this might be an issue unless witnesses also add support for requesting previously co-signed checkpoints.

haydentherapper commented:
> I think the one case where I'm not confident is where the v2 rekor instance is added to signing config:
> 
> * before any clients are updated this is straight forward: clients keep signing with v1, nothing breaks
> * when client A is upgraded to support v2, the docs seem to imply that it should then only use the v2 instance when the signingconfig includes one  (correct?)
> * this leads to a client incompatibility if other clients are not able to verify rekor v2 at this point yet
> 
> Basically we may need to confirm verification support across the ecosystem before we make signingconfig changes where a rekor v2 instance is enabled

cc @loosebazooka for something to discuss at the client meeting

This is a consequence of making a TLE v2 struct unfortunately, though even if we didn't, not all clients support signed timestamps so a TLE without an integrated timestamp wouldn't be verifiable anyways - there would have been incompatibility regardless. There will be a period in which some clients (-ruby and -js) will not be able to verify bundles produced by other clients.

I'd like to chat more with GitHub to understand how much cross-compatibility we need between -js and -go for artifact attestations, if any is actually needed.

jku commented on CLIENTS.md:
```diff
@@ -0,0 +1,369 @@
+bundle, and clients use that log ID to lookup the correct log public key
+in the TrustedRoot to verify the bundle.
+
+Rekor will no longer include log IDs in the response, and instead clients
+should use the checkpoint key ID as specified in the
+[C2SP spec](https://github.com/C2SP/C2SP/blob/main/signed-note.md#signatures)
+to lookup the correct log public key to verify a bundle. Each transparency
```
I think this might not be clear enough like this. How exactly do clients calculate the checkpoint key id for an entry -- is entrys checkpoint origin the "key name"?
 
 is there an example implementation of this?

haydentherapper replied:
Added selection algorithms for this, thanks. The `TransparencyLogInstance` proto has been updated as well to document this.

tl;dr either 1) compare the bundle's `log_id` to the trust root log instance's `checkpoint_key_id` or `log_id` (what most clients are doing already), 2) enumerate all log instances and compare the checkpoint's key ID from the signature line to the first 4 bytes of the trust root log instance's `checkpoint_key_id`/`log_id` (which is what sigstore-python is doing), or 3) compute the log instance's checkpoint key ID using the instance's public key and origin and compare (this is Hard Mode, and I doubt any client will)

ramonpetgrave64 commented on CLIENTS.md:
````diff
@@ -0,0 +1,369 @@
+}
+```
+
+The response will be a
+[`TransparencyLogEntry` message](https://github.com/sigstore/protobuf-specs/blob/5296f13d62e7fad428581d969f664c30cc52f549/protos/sigstore_rekor.proto#L94),
+which should be persisted in a bundle. Clients no longer need to transform the
+Rekor response into a `TLE` message to store in the bundle.
````
It would be helpful to show the shapes of the old and new responses.

haydentherapper commented:
Note to self: Split into changes for signingconfig, trustroot, and bundle. Add KindVersion and log ID (which is not hex-encoded, unlike rekor v1)

cmurphy commented on CLIENTS.md:
```diff
@@ -0,0 +1,369 @@
+
+Clients need to increase request timeouts to at least 10 seconds.
+
+Rekor now batches uploads so that checkpoints are published less frequently.
```
Not that this matters to clients but this is technically not true, the checkpoint publishing interval is not related to the batch interval.

haydentherapper replied:
Thanks, removed the mention of batching and clarified that this is just about blocking until a checkpoint has been published.

cmurphy commented on CLIENTS.md:
```diff
@@ -0,0 +1,369 @@
+of a `log_id`.
+
+For more information, log IDs are not necessarily unique identifiers for
+a log, since a log may reuse its public key among instances. Additionally,
```
Why would we not just create a new key for the new instance?

haydentherapper replied:
Added a line that for the public deployment, we always will. Private deployments may reuse keys.

cmurphy commented on CLIENTS.md:
```diff
@@ -0,0 +1,369 @@
+For more information, log IDs are not necessarily unique identifiers for
+a log, since a log may reuse its public key among instances. Additionally,
+the log origin is not necessarily a unique identifier, because multiple logs
+may be hosted by one origin. Even the combination of both is not necessarily
```
Each shard should be uniquely named

haydentherapper replied:
Thanks, added a mention that this SHOULD NOT happen. Technically, a deployment can reuse the key and/or origin, which the checkpoint key ID calculation takes into account.

cmurphy commented on CLIENTS.md:
```diff
@@ -0,0 +1,369 @@
+    // it supports.
+    "rekorTlogUrls": [
+        {
+            "url": "https://rekor-2025-1.sigstore.dev",
```
```suggestion
            "url": "https://log2025-01.rekor.sigstore.dev",
```

cmurphy replied:
```suggestion
            "url": "https://log2025-01.rekor.sigstore.dev",
```

ramonpetgrave64 commented on CLIENTS.md:
```diff
@@ -367,3 +367,59 @@ be distributed by Sigstore's TUF root.
+      "envelope": "rekor-local\n1\nm5JVGx4ESzbU2kFUPgYQ9adCA7e7mjnwQRluEmGJbHU=\n\n— rekor-local 2AtEIIwnbtxrneJ7L1lQebfBRl7TxK84DTmx+kcZi7A25cBDgESI23f9ylThAlOireJ7U+H8eZF/4kJQcn9o5Qt8mQU=\n"
+    }
+  },
+  "canonicalizedBody": "eyJhcGlWZXJzaW9uIjoiMC4wLjIiLCJraW5kIjoiaGFzaGVkcmVrb3JkIiwic3BlYyI6eyJoYXNoZWRSZWtvcmRWMF8wXzIiOnsiZGF0YSI6eyJhbGdvcml0aG0iOiJTSEEyXzI1NiIsImRpZ2VzdCI6ImR5ajRlZG5ZSGpONC96c2pqQmVlTGFoUzlzbHA5N1o2N0xUQVZ4anJqWHc9In0sInNpZ25hdHVyZSI6eyJjb250ZW50IjoiTUVRQ0lCK1lQYTlvM1NOMHNRNHVkdUdmK21aeHdGZk9oRlowQ2d5K3A3VnQxbzJTQWlBUEZESHFPQUpMWW12dENXT3NEeU5ZMUg0VjN6bTRORURZczNOeXZIaDFQZz0 ...
```
Also show the decoded canonicalized body.

haydentherapper replied:
There is now a section on parsing the canonicalized body.

haydentherapper commented:
Thanks everyone for comments! I've addressed all comments now. The primary changes are adding a section on entry parsing, some details about bundle-related changes, more info about log instance lookup in the trusted root, and a section on testing against a real instance using `curl`. The SigningConfig documentation has also been updated in the protobuf-specs repo, it's linked here.

jku reviewed (approved):
This looks great, no notes from me after reading through it twice. Thanks for polishing.

ramonpetgrave64 commented on CLIENTS.md:
```diff
@@ -0,0 +1,631 @@
+
+### Handling Longer Requests
+
+Clients need to increase request timeouts to at least 10 seconds.
```
```suggestion
Clients need to increase request timeouts to at least 10 seconds when creating entries.
```

haydentherapper replied:
Added, and I've doubled the timeout recommendation to 20 seconds.
